File: desktop/macos/Sources/NexusDesktop/Sidecar/SidecarBundleLocator.swift
import Foundation

struct SidecarBundleLocator {
  let projectRoot: URL?
  let webDistURL: URL
  let appRootURL: URL
  let command: String
  let arguments: [String]
  let workingDirectory: URL

  static func resolve() throws -> SidecarBundleLocator {
    if let bundled = try resolveBundled() {
      return bundled
    }
    return try resolveDevelopment()
  }

  private static func resolveBundled() throws -> SidecarBundleLocator? {
    guard let resourceURL = Bundle.main.resourceURL else {
      return nil
    }
    let webDistURL = resourceURL.appendingPathComponent("Web", isDirectory: true)
    let serverURL = resourceURL
      .deletingLastPathComponent()
      .appendingPathComponent("MacOS", isDirectory: true)
      .appendingPathComponent("nexus-server")
    guard FileManager.default.fileExists(atPath: webDistURL.appendingPathComponent("index.html").path) else {
      return nil
    }
    guard FileManager.default.isExecutableFile(atPath: serverURL.path) else {
      return nil
    }
    return SidecarBundleLocator(
      projectRoot: nil,
      webDistURL: webDistURL,
      appRootURL: resourceURL,
      command: serverURL.path,
      arguments: [],
      workingDirectory: resourceURL
    )
  }

  private static func resolveDevelopment() throws -> SidecarBundleLocator {
    guard let root = findProjectRoot() else {
      throw DesktopShellError.projectRootNotFound
    }
    let webDistURL = root.appendingPathComponent("web/dist", isDirectory: true)
    guard FileManager.default.fileExists(atPath: webDistURL.appendingPathComponent("index.html").path) else {
      throw DesktopShellError.webDistNotFound(webDistURL.path)
    }
    return SidecarBundleLocator(
      projectRoot: root,
      webDistURL: webDistURL,
      appRootURL: root,
      command: "/usr/bin/env",
      arguments: ["go", "run", "./cmd/nexus-server"],
      workingDirectory: root
    )
  }

  private static func findProjectRoot() -> URL? {
    let candidates = [
      URL(fileURLWithPath: FileManager.default.currentDirectoryPath),
      URL(fileURLWithPath: #filePath),
    ]
    for candidate in candidates {
      var current = candidate.hasDirectoryPath ? candidate : candidate.deletingLastPathComponent()
      while current.path != current.deletingLastPathComponent().path {
        let goMod = current.appendingPathComponent("go.mod")
        let webIndex = current.appendingPathComponent("web/index.html")
        if FileManager.default.fileExists(atPath: goMod.path),
           FileManager.default.fileExists(atPath: webIndex.path) {
          return current
        }
        current.deleteLastPathComponent()
      }
    }
    return nil
  }
}

File: desktop/windows/Nexus.Desktop/Sidecar/SidecarBundleLocator.cs
using System.IO;

namespace Nexus.Desktop.Sidecar;

internal sealed record SidecarBundle(
    string AppRoot,
    string WebDistDirectory,
    string Command,
    string Arguments,
    string WorkingDirectory,
    bool IsDevelopment);

internal static class SidecarBundleLocator
{
    public static SidecarBundle Resolve()
    {
        SidecarBundle? bundled = ResolveBundled();
        if (bundled is not null)
        {
            return bundled;
        }

        return ResolveDevelopment();
    }

    private static SidecarBundle? ResolveBundled()
    {
        string baseDirectory = AppContext.BaseDirectory;
        string webDist = Path.Combine(baseDirectory, "Resources", "Web");
        string sidecar = Path.Combine(baseDirectory, "Resources", "nexus-server.exe");
        if (File.Exists(Path.Combine(webDist, "app.html")) && File.Exists(sidecar))
        {
            return new SidecarBundle(
                AppRoot: Path.Combine(baseDirectory, "Resources"),
                WebDistDirectory: webDist,
                Command: sidecar,
                Arguments: string.Empty,
                WorkingDirectory: Path.Combine(baseDirectory, "Resources"),
                IsDevelopment: false);
        }

        return null;
    }

    private static SidecarBundle ResolveDevelopment()
    {
        string? root = FindProjectRoot();
        if (root is null)
        {
            throw new InvalidOperationException("未找到 Nexus 仓库根目录。");
        }

        string webDist = Path.Combine(root, "web", "dist");
        if (!File.Exists(Path.Combine(webDist, "app.html")))
        {
            throw new InvalidOperationException($"未找到 web/dist/app.html：{webDist}");
        }

        return new SidecarBundle(
            AppRoot: root,
            WebDistDirectory: webDist,
            Command: "go",
            Arguments: "run ./cmd/nexus-server",
            WorkingDirectory: root,
            IsDevelopment: true);
    }

    private static string? FindProjectRoot()
    {
        string current = Directory.GetCurrentDirectory();
        while (!string.IsNullOrWhiteSpace(current))
        {
            if (File.Exists(Path.Combine(current, "go.mod")) && File.Exists(Path.Combine(current, "web", "index.html")))
            {
                return current;
            }

            DirectoryInfo? parent = Directory.GetParent(current);
            if (parent is null)
            {
                return null;
            }
            current = parent.FullName;
        }

        return null;
    }
}
